src/lang/checkers/SanityChecker.ml

```diff
@@ -69,12 +77,104 @@ module ScillaSanityChecker
         recurser ilist []
     in
 
+    let find_accept_groups (stmts : stmt_annot list) : loc list list =
+      let rec walk (seen : loc list list) (stmts : stmt_annot list) : loc list list =
+        (* Walk the syntax tree looking for code paths containing one
+           or more accept statements.
+
+           seen is a List with an item for each code path which has
+           already seen at least one accept statement before reaching
+           the point in the syntax tree just before stmts.  Each item
+           in the List is List of all the accepts seen along that path
+           to this point.
+
+           Returns an updated version of the List of Lists which
+           includes any accepts seen in stmts.
+
+           Note that any given code path remains untracked until an
+           accept is discovered in it.  This keeps the seen data
+           nested List to a minimal size. *)
+        List.fold_left
+          (fun (seen2 : loc list list) (stmt : stmt_annot) ->
+            let loc = (stmt_loc stmt) in
+            match fst stmt with
+            | AcceptPayment ->
+               (match seen2 with
+                (* If we didn't see any accept statements before reaching
+                 * this point, register this as the first one. *)
+                | [] -> [[loc]]
+                (* Otherwise add this accept statement to the list of accepts
+                 * already seen on each code path reaching this point. *)
+                | _ -> List.map (fun accepts -> loc :: accepts) seen2)
+            | MatchStmt (_ident, branches) ->
+               (* For each branch in the match statement we have a
+                    new code path to "multiply" with the code paths
+                    which already reached this point, so walk each
+                    branch and build all the results into a new list.
+                *)
+               List.fold_left
+                 (fun seen3 (_pattern, branchstmts) ->
+                   match walk seen2 branchstmts with
+                   | [] -> seen3
+                   | seen4 -> seen3 @ seen4)
+                 [] branches
+            | _ -> seen2
+          ) (seen : loc list list) (stmts : stmt_annot list)
+      in
+        walk [] stmts
+    in
+
+    let check_accepts (contr : contract) =
+      let check_transition_accepts (transition : transition) =
+        let transition_accept_groups = List.map List.rev (find_accept_groups transition.tbody) in
+
+        let dup_accept_warning (group : loc list) : unit =
+          (warn1
+             (Core.sprintf "transition %s had a potential code path with duplicate accept statements:\n"
+                (get_id transition.tname) ^
+                String.concat ""
+                  (List.map
+                     (fun loc -> Core.sprintf "  Accept at %s\n" (get_loc_str loc))
```

Comment:
I wonder if the error messages could not be improved.

We don't expect there to be many `accept` statements in a code path, but maybe the best thing to do is use `warn2` which takes both a start location and an end location. You can then pass the location of either the first and second `accept`, or the location of the first and the last `accept` in the code path.

This may also simplify the analysis, since you won't need to keep track of more than 2 locations in each code path.

Reply:
I think I previously tried to use `warn2` but then couldn't see any way of getting the end location.  Having said that, in this case `accept` is always 6 characters so I guess adding 6 to the start column would be good enough.

But IIUC here you are suggesting something different, i.e. to make the start location point to the beginning of one `accept`, and the end location point to the beginning of another - right?  I can think of two potential objections to that:

1. It would seem slightly inconsistent with all other warnings/errors where the expression causing the warning/error is contained within the start and end locations.  IIUC with your suggestion there would be a culprit expression *starting* at the warning's end location.

2. I deliberately decided to collect *all* instances of `accept` within each code path rather than just the minimum of 2 required to show a warning, so that if there were more than 2, the Scilla developer would be able to check (and potentially fix) them all in one go.  OTOH if the warning only shows 2 but there are more than 2, the developer would have to eliminate one of those in order to check (and potentially fix) the others.  And maybe that elimination would actually need to be reverted after checking the others, since the `accept` eliminated might not be a true duplicate for reasons which a human can deduce but which static analysis cannot (yet) infer (as per the example in this PR's description).  Worse, the developer wouldn't even be aware that there were more than 2 potential `accepts` in the same code path, so it might be that a non-duplicate warning could actually prevent any visible warning of a genuine duplicate.

The first of these two is more of an aesthetic inconsistency which could quite easily be ignored; however the second seems to infringe on the correctness of the check, which concerns me much more.

Having said that, I'm hopelessly new to all of this, so of course I defer to your expertise ;-)

Reply:
Ad. 1: That is indeed my suggestion. I don't see a huge problem in having a the end position of a warning be at the start of the violating `accept` statement, even if it is slightly inconsistent.

Ad. 2: When the typechecker discovers an error in a transition, we bail out of that transition regardless of whether there may be more (related or unrelated) type errors in the same transition, so it's not entirely inconsistent to only report two of the violating `accept`s.

The issue is not only with the console output, but also with the integration with Savant IDE (https://ide-next.zilliqa.com/) or other IDEs. If the error position is given just as one of the violating statements, then the IDE won't be able to highlight the other violating statement unless it attempts to parse the error message string.

On the other hand, if it is given the positions of the first and last violating statements, then a naïve IDE would highlight the entire code block from the first to the last violating statement, and this would most likely include all the other violating statements in the transition.

Reply:
Thanks for the great explanations - I have a much better understanding now of where you're coming from.

Ad. 1: We could even have the end position be 6 characters after the start of the last `accept` in the warned codepath, so that the warning range includes every `accept` which it refers to.

Ad. 2: Points taken about the (lack of) inconsistency and the IDE highlighting.

I think the best approach which takes all these points into account is my suggestion immediately above in this comment (Ad. 1) coupled with a warning message which includes *all* potential duplicate `accepts` in the code path.  As [previously mentioned](https://github.com/Zilliqa/scilla/pull/409#discussion_r254678601), this last detail is necessary in order to maintain the correctness of the check, since if only 2 `accept` statements are mentioned in the warning and the developer decides that those are not worth worrying about, they might surmise that there are no other `accepts` in the same code path, which might not be the case.

Reply:
OK, I've now done this.